src/crimson/os/seastore/device.h

```diff
@@ -0,0 +1,118 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:nil -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <memory>
+
+#include "include/buffer_fwd.h"
+
+#include "crimson/common/errorator.h"
+#include "crimson/os/seastore/seastore_types.h"
+
+namespace crimson::os::seastore {
+
+using magic_t = uint64_t;
+
+struct device_spec_t{
+  magic_t magic;
+  device_type_t dtype;
+  device_id_t id;
+  DENC(device_spec_t, v, p) {
+    DENC_START(1, 1, p);
+    denc(v.magic, p);
+    denc(v.dtype, p);
+    denc(v.id, p);
+    DENC_FINISH(p);
+  }
+};
+
+std::ostream& operator<<(std::ostream&, const device_spec_t&);
+
+using secondary_device_set_t =
+  std::map<device_id_t, device_spec_t>;
+
+struct device_config_t {
+  bool major_dev = false;
+  magic_t magic = 0;
+  device_type_t dtype = device_type_t::NONE;
+  device_id_t device_id = DEVICE_ID_NULL;
+  seastore_meta_t meta;
+  secondary_device_set_t secondary_devices;
+};
+
+std::ostream& operator<<(std::ostream&, const device_config_t&);
+
+class Device;
+using DeviceRef = std::unique_ptr<Device>;
+
+/**
+ * Device
+ *
+ * Represents a general device regardless of the underlying medium.
+ */
+class Device {
```

Comment:
This interface seems read-specific.  How about DeviceReader?

Reply:
This `Device` class is added with more interfaces to work with the general EPM and SeaStore classes in the follow-up commits.

Should I split the `Device` into `Device` and `DeviceReader`?

Reply:
Mmm, more like a parent class for SegmentManager and RandomBlockManager, I suppose.  Probably fine as Device.

Reply:
Sorry, changed my mind after my first comment, fine as is.